File: frontend/miniprogram/src/api/selfExam.ts
import { request } from '@/utils/request'
import type { ExamResultOut } from '@/types/api'

export interface SelfExamQuota {
  is_promax: boolean
  used: number
  limit: number
  remaining: number
}

export interface SelfExamQuestion {
  id: string
  question_type: string
  stem: string
  options?: string[] | null
  difficulty?: number | null
}

export interface SelfExamOut {
  id: string
  status: string
  time_limit_sec: number
  weak_kps: string[]
  questions: SelfExamQuestion[]
  total?: number | null
  correct_count?: number | null
  accuracy?: number | null
  created_at: string
}

export interface SelfExamBrief {
  id: string
  status: string
  total?: number | null
  correct_count?: number | null
  accuracy?: number | null
  created_at: string
}

export function getSelfExamQuota(): Promise<SelfExamQuota> {
  return request<SelfExamQuota>('/api/v1/self-exam/quota', { method: 'GET' })
}

export function generateSelfExam(): Promise<SelfExamOut> {
  return request<SelfExamOut>('/api/v1/self-exam/generate', { method: 'POST' })
}

export function getSelfExam(id: string): Promise<SelfExamOut> {
  return request<SelfExamOut>(`/api/v1/self-exam/${id}`, { method: 'GET' })
}

export function submitSelfExam(
  id: string, answers: { question_id: string; user_answer: string }[],
): Promise<{ result: ExamResultOut; exam: SelfExamBrief }> {
  return request(`/api/v1/self-exam/${id}/submit`, { method: 'POST', data: { answers } })
}

export function getSelfExamHistory(): Promise<SelfExamBrief[]> {
  return request<SelfExamBrief[]>('/api/v1/self-exam/history', { method: 'GET' })
}

File: backend/app/schemas/self_exam.py
"""ProMax 自助出卷 schemas（5C）。"""
from __future__ import annotations

import datetime as dt
import uuid

from pydantic import BaseModel

from app.schemas.questions import ExamResultOut, PracticeAttemptIn


class SelfExamQuota(BaseModel):
    is_promax: bool
    used: int
    limit: int
    remaining: int


class SelfExamQuestion(BaseModel):
    id: str
    question_type: str
    stem: str
    options: list[str] | None = None
    difficulty: int | None = None


class SelfExamOut(BaseModel):
    """答题页用：含题目快照（不含答案）+ 限时。"""
    id: uuid.UUID
    status: str
    time_limit_sec: int
    weak_kps: list[str] = []
    questions: list[SelfExamQuestion] = []
    total: int | None = None
    correct_count: int | None = None
    accuracy: float | None = None
    created_at: dt.datetime


class SelfExamBrief(BaseModel):
    id: uuid.UUID
    status: str
    total: int | None = None
    correct_count: int | None = None
    accuracy: float | None = None
    created_at: dt.datetime


class SelfExamSubmitIn(BaseModel):
    answers: list[PracticeAttemptIn]


class SelfExamSubmitResult(BaseModel):
    result: ExamResultOut
    exam: SelfExamBrief
